Source/Tools/LinuxEmulation/LinuxSyscalls/Utils/Threads.cpp

```diff
@@ -61,6 +82,32 @@ namespace FEX::LinuxEmulation::Threads {
     AddStackToDeadPool(Ptr);
   }
 
+  [[noreturn]]
+  void DeallocateStackObjectAndExit(void *Ptr, int Status) {
+    RemoveStackFromLivePool(Ptr);
+    auto ReadyToBeReaped = AddStackToDeadPool(Ptr);
+    *ReadyToBeReaped = true;
```

Comment:
There's no other logic in here, and the code reading `ReadyToBeReaped` is protected by a mutex. Can't we move setting the boolean into AddStackToDeadPool and avoid the obscure bool pointer return altogether hence? This could be optional behavior controlled by a function parameter. (This might also make it clearer why ReadyToBeReaped must be set here but not in the other call site of AddStackToDeadPool.)

Reply:
We can move the ReadyToBeReaped store in to `AddStackToDeadPool` if we continue passing the status all the way to that function so we can call exit there.
At the point of `ReadyToBeReaped` being set to true, we no longer have ownership of the stack. We can't return from a function, we can't call a function, we must do the syscall immediately.
The mutex guarding this doesn't matter at all, we are dancing around the thread no longer having a stack.

Reply:
The stack ownership is released when setting ReadyToBeReaped indeed, but this ownership isn't re-assigned until `AllocateStackObject` observes `ReadyToBeReaped==1`. This observation can't happen as long as `DeadStackPoolMutex` is locked. Until then, accessing the stack should still be safe.

So instead of the boolean pointer dance, why not keep `DeadStackPoolMutex` locked until after we return from `AddStackDeadPool`? This could be achieved for example by moving the mutex locking out of `AddStackToDeadPool` and locking/unlocking the mutex here manually. (To preserve current behavior elsewhere, the function could be renamed to "AddStackToDeadPoolInternal" and a helper function could be added that behaves like the current code).

Reply:
As discussed externally, this won't work because unlocking the mutex itself requires a stack. (Leaving discussion open for visibility.)